### PR TITLE
[ Misc ] Refactor MoE to isolate Fp8 From Mixtral

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
MOTIVATION:
* Prior to this PR, MoE layer + weight loading logic is repeated for all models
* Prior to this PR, fp8 logic for MoE is coupled with Mixtral and unsupported for other MoE Models as a result
* Prior to this PR, difficult to add other quantization methods without significant duplicated code in model files, which is very hard to maintain (e.g. made it hard to land #2761)

THIS PR:
* Creates concept of `FusedMoEMethodBase`, `UnquantizedFusedMoEMethod`, `Fp8FusedMoeMethod`
* Creates concept of `FusedMoE` layer (akin to `MergedColumnParallel` + `RowParallel`
* Confirm correctness with lm-eval harness
* Confirm able to use `FusedMoELinear` with Qwen2Moe

FOLLOW UP PR:
* Support Fp8 checkpoints for models other than Mixtral: (weight loading for scales is currently overfit to Mixtral)
* Convert other MoE models to use `FusedMoELinear`: to enable Fp8 for those cases
* Finally integrate GPTQ / AWQ MoE kernels (#2761)

ACCURACY:
- `Mixtral`

```bash
MODEL="mistralai/Mixtral-8x7B-Instruct-v0.1"
lm_eval --model vllm \
  --model_args pretrained=$MODEL,tensor_parallel_size=4,distributed_executor_backend="ray" \
  --tasks gsm8k --num_fewshot 5 --limit 1000 --batch_size "auto"

main:
|Tasks|Version|     Filter     |n-shot|  Metric   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|----:|---|-----:|
|gsm8k|      3|strict-match    |     5|exact_match|0.643|±  |0.0152|
|     |       |flexible-extract|     5|exact_match|0.648|±  |0.0151|

branch:
|Tasks|Version|     Filter     |n-shot|  Metric   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|----:|---|-----:|
|gsm8k|      3|strict-match    |     5|exact_match|0.643|±  |0.0152|
|     |       |flexible-extract|     5|exact_match|0.648|±  |0.0151|
```

- `Qwen2 MoE`

```bash
MODEL="Qwen/Qwen2-57B-A14B-Instruct"
lm_eval --model vllm \
  --model_args pretrained=$MODEL,tensor_parallel_size=4,distributed_executor_backend="ray" \
  --tasks gsm8k --num_fewshot 5 --limit 1000 --batch_size "auto"

main:
Tasks|Version|     Filter     |n-shot|  Metric   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|----:|---|-----:|
|gsm8k|      3|strict-match    |     5|exact_match|0.812|±  |0.0124|
|     |       |flexible-extract|     5|exact_match|0.834|±  |0.0118|


- branch:
|Tasks|Version|     Filter     |n-shot|  Metric   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|----:|---|-----:|
|gsm8k|      3|strict-match    |     5|exact_match|0.810|±  |0.0124|
|     |       |flexible-extract|     5|exact_match|0.834|±  |0.0118|
```

Until we have the large models re-added to the CI, I run the following script offline:

```bash
TP_SIZE=4
FEWSHOT=5
LIMIT=250
BATCH_SIZE="auto"

# Mixtral
MODEL="mistralai/Mixtral-8x7B-Instruct-v0.1"
lm_eval --model vllm \
  --model_args pretrained=$MODEL,tensor_parallel_size=$TP_SIZE,distributed_executor_backend="ray" \
  --tasks gsm8k --num_fewshot $FEWSHOT --limit $LIMIT --batch_size $BATCH_SIZE

# Mixtral - Quantized in place
MODEL="mistralai/Mixtral-8x7B-Instruct-v0.1"
lm_eval --model vllm \
  --model_args pretrained=$MODEL,tensor_parallel_size=$TP_SIZE,distributed_executor_backend="ray",quantization="fp8" \
  --tasks gsm8k --num_fewshot $FEWSHOT --limit $LIMIT --batch_size $BATCH_SIZE

# Mixtral - Quantized
MODEL="neuralmagic/Mixtral-8x7B-Instruct-v0.1-FP8"
lm_eval --model vllm \
  --model_args pretrained=$MODEL,tensor_parallel_size=$TP_SIZE,distributed_executor_backend="ray",quantization="fp8" \
  --tasks gsm8k --num_fewshot $FEWSHOT --limit $LIMIT --batch_size $BATCH_SIZE

# Qwen
MODEL="Qwen/Qwen2-57B-A14B-Instruct"
lm_eval --model vllm \
  --model_args pretrained=$MODEL,tensor_parallel_size=$TP_SIZE,distributed_executor_backend="ray" \
  --tasks gsm8k --num_fewshot $FEWSHOT --limit $LIMIT --batch_size $BATCH_SIZE

# Qwen - Quantized in place
MODEL="Qwen/Qwen2-57B-A14B-Instruct"
lm_eval --model vllm \
  --model_args pretrained=$MODEL,tensor_parallel_size=$TP_SIZE,distributed_executor_backend="ray",quantization="fp8" \
  --tasks gsm8k --num_fewshot $FEWSHOT --limit $LIMIT --batch_size $BATCH_SIZE

# Llama3
MODEL="meta-llama/Meta-Llama-3-8B-Instruct"
lm_eval --model vllm \
  --model_args pretrained=$MODEL,tensor_parallel_size=$TP_SIZE,distributed_executor_backend="ray" \
  --tasks gsm8k --num_fewshot $FEWSHOT --limit $LIMIT --batch_size $BATCH_SIZE

# Llama-3 Quantized in place
MODEL="meta-llama/Meta-Llama-3-8B-Instruct"
lm_eval --model vllm \
  --model_args pretrained=$MODEL,tensor_parallel_size=$TP_SIZE,distributed_executor_backend="ray",quantization="fp8" \
  --tasks gsm8k --num_fewshot $FEWSHOT --limit $LIMIT --batch_size $BATCH_SIZE

# Llama-3 - Quantized
MODEL="neuralmagic/Meta-Llama-3-8B-Instruct-FP8"
lm_eval --model vllm \
  --model_args pretrained=$MODEL,tensor_parallel_size=$TP_SIZE,distributed_executor_backend="ray" \
  --tasks gsm8k --num_fewshot $FEWSHOT --limit $LIMIT --batch_size $BATCH_SIZE
```

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/layers/linear.py [hunk 1]
```diff
@@ -97,6 +98,21 @@ def apply(self,
         Expects create_weights to have been called before on the layer."""
         raise NotImplementedError
 
+    def create_weights_moe(self, layer: torch.nn.Module, num_experts: int,
+                           hidden_size: int, intermediate_size: int,
+                           params_dtype: torch.dtype, **extra_weight_attrs):
+        raise NotImplementedError(
+            "Moe models are not yet supported for your LinearMethod")
```
Comment (comaniac) [inline vllm/model_executor/layers/linear.py:105]: Can we use `self.__class__.name` here to replace "your LinearMethod"?

File: vllm/model_executor/layers/linear.py [hunk 2]
```diff
@@ -97,6 +98,21 @@ def apply(self,
...
+
+    def apply_moe(self,
+                  layer: torch.nn.Module,
+                  x: torch.Tensor,
+                  router_logits: torch.Tensor,
+                  top_k: int,
+                  renormalize: bool = True) -> torch.Tensor:
+        raise NotImplementedError(
+            "Moe models are not yet supported for your LinearMethod")
```
Comment (comaniac) [inline vllm/model_executor/layers/linear.py:114]: This API is a bit weird. Given that our MoE actually has nothing to do with Linear layers, it might be better to have a separate `moe.py` with `class MoE`, so that we can keep unified APIs `create_weights` and `apply`.

File: vllm/model_executor/layers/linear.py (hunk 1 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/linear.py:105]: Yes

File: vllm/model_executor/layers/linear.py (hunk 2 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/linear.py:114]: Sounds good - I almost did it that way. Will restructure

File: .buildkite/lm-eval-harness/run-lm-eval-gsm-vllm-baseline.sh [hunk 3]
```diff
@@ -46,6 +46,6 @@ while getopts "m:b:l:f:t:" OPT; do
 done
 
 lm_eval --model vllm \
-  --model_args pretrained=$MODEL,tensor_parallel_size=$TP_SIZE \
+  --model_args pretrained=$MODEL,tensor_parallel_size=$TP_SIZE,distributed_executor_backend="ray" \
```
Comment (mgoin) [inline .buildkite/lm-eval-harness/run-lm-eval-gsm-vllm-baseline.sh:49]: I thought ray was used by default if it was installed, does it need to be specified manually now?

File: vllm/model_executor/layers/linear.py [hunk 4]
```diff
@@ -775,3 +828,105 @@ def extra_repr(self) -> str:
         s += f", tp_size={self.tp_size}"
         s += f", reduce_results={self.reduce_results}"
         return s
+
+
+class FusedMoELinear(torch.nn.Module):
+
```
Comment (mgoin) [inline vllm/model_executor/layers/linear.py:834]: It's worth putting a docstring here for model implementers clarifying what w1, w2, w3 usually map to and what structure this layer encapsulates for MoEs in vLLM

File: vllm/model_executor/layers/quantization/fp8.py [hunk 5]
```diff
@@ -224,9 +293,96 @@ def process_weights_after_loading(self, layer: Module) -> None:
...
+                start = 0
+                for shard_id in range(2):
+                    dq_weight = per_tensor_dequantize(
+                        layer.w13_weight[expert_id][start:start +
+                                                    shard_size, :],
+                        layer.w13_scale[expert_id][shard_id])
+                    layer.w13_weight[expert_id][
+                        start:start + shard_size, :] = per_tensor_quantize(
+                            dq_weight, max_w13_scales[expert_id])
+                    start += shard_size
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/fp8.py:366]: This could be a bit more clear with explicit start and end
```suggestion
            for expert_id in range(layer.num_experts):
                for shard_id in range(2):
                    start = shard_id * shard_size
                    end = start + shard_size
                    dq_weight = per_tensor_dequantize(
                        layer.w13_weight[expert_id][start:end, :],
                        layer.w13_scale[expert_id][shard_id])
                    layer.w13_weight[expert_id][
                        start:end, :] = per_tensor_quantize(
                            dq_weight, max_w13_scales[expert_id])
```

File: vllm/model_executor/layers/quantization/fp8.py [hunk 6]
```diff
@@ -224,9 +293,96 @@ def process_weights_after_loading(self, layer: Module) -> None:
...
+        # MoE kernels require single activation scale and single weight
+        # scale for w13 per expert.
+        else:
+            # Fp8 moe kernels require a single activation scale.
+            # We take the max of all the scales in case they differ.
+            if self.quant_config.activation_scheme == "static":
+                if layer.a13_scale is None or layer.a2_scale is None:
+                    raise ValueError(
+                        "QuantConfig has static quantization, but found "
+                        "activation scales are None.")
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/fp8.py:339]: Would be nice to print both so you can see if just one is empty

File: vllm/model_executor/layers/quantization/fp8.py [hunk 7]
```diff
@@ -224,9 +293,96 @@ def process_weights_after_loading(self, layer: Module) -> None:
...
+                dtype=torch.float32,
+                device=w13_weight.device),
+                                                 requires_grad=False)
+            for expert in range(layer.num_experts):
+                w13_weight[expert, :, :], layer.w13_scale[
+                    expert] = ops.scaled_fp8_quant(
+                        layer.w13_weight.data[expert, :, :])
+                w2_weight[expert, :, :], layer.w2_scale[
+                    expert] = ops.scaled_fp8_quant(
+                        layer.w2_weight.data[expert, :, :])
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/fp8.py:322]: nit: I don't think you need the colons here

File: vllm/model_executor/models/qwen2_moe.py [hunk 8]
```diff
@@ -426,21 +397,36 @@ def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
...
+            for expert_id in range(self.config.num_experts) for shard_id,
+            weight_name in enumerate(["gate_proj", "down_proj", "up_proj"])
+        ]
+
         params_dict = dict(self.named_parameters())
         for name, loaded_weight in weights:
             if "rotary_emb.inv_freq" in name:
                 continue
             for (param_name, weight_name, shard_id) in stacked_params_mapping:
+                # Skip non-stacked and experts (experts handled below).
```
Comment (mgoin) [inline vllm/model_executor/models/qwen2_moe.py:415]:
```suggestion
                # Skip non-stacked layers and experts (experts handled below).
```

File: vllm/model_executor/layers/linear.py [hunk 9]
```diff
@@ -775,3 +828,105 @@ def extra_repr(self) -> str:
...
+    def __init__(
+        self,
+        num_experts: int,
+        top_k: int,
+        hidden_size: int,
+        intermediate_size: int,
+        params_dtype: Optional[torch.dtype] = None,
+        reduce_results: bool = False,
```
Comment (mgoin) [inline vllm/model_executor/layers/linear.py:842]: I think reduce results should be True by default to deal with TP by default

File: .buildkite/lm-eval-harness/run-lm-eval-gsm-vllm-baseline.sh (hunk 3 above)
Comment (robertgshaw2-neuralmagic) [reply .buildkite/lm-eval-harness/run-lm-eval-gsm-vllm-baseline.sh:49]: `mp` is default

Was having issues using this inside lm-eval-harness with `mp` unrelated to this PR

File: vllm/model_executor/layers/quantization/fp8.py (hunk 7 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fp8.py:389]: Im going to leave it since this i how it was originally done in mixtral

File: vllm/model_executor/layers/linear.py (hunk 4 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/linear.py]: Done

File: vllm/model_executor/layers/quantization/fp8.py (hunk 5 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/fp8.py:431]: im going to leave it because this is how it was done in mixtral. We can do a follow up for just changing this piece

**Comment by robertgshaw2-neuralmagic** [comment]
@comaniac ready for re-review

**Comment by robertgshaw2-neuralmagic** [comment]
This should be good to go. 

* Note that Qwen and Mixtral run in the CI at fp16
* I ran the accuracy results for fp8 offline

**Review by comaniac — APPROVED** [review]
LGTM for the refactoring. Should be good to go when CI is green.
